src/print.c

```diff
@@ -245,6 +259,189 @@ launch_preview (const char *filename,
+  g_ptr_array_add (args, g_strdup_printf ("--raw=%d", page));
+  g_ptr_array_add (args, NULL);
+  process = g_subprocess_newv ((const gchar * const *)args->pdata, G_SUBPROCESS_FLAGS_STDOUT_PIPE, NULL);
+
+  if (process)
+  {
+    stream = g_subprocess_get_stdout_pipe (process);
+    if (stream)
+    {
+      size_t max_size = w * h * 4;
+      data = (unsigned char *) g_malloc (max_size);
+      while ((read = g_input_stream_read (stream, data + total_read, 1024, NULL, NULL)) > 0 && max_size > total_read)
+      {
+         total_read += read;
+      }
+    }
+  }
+  return data;
+}
+
+static void
+pdf_begin_print (GtkPrintOperation *op,
+                 GtkPrintContext   *ctx,
+                 char              *filename)
+{
+    gtk_print_operation_set_n_pages (op, pdf_get_actions_page (filename, PDF_NUM_PAGES, 0));
+}
+
+static void
+pdf_draw_page(GtkPrintOperation *op,
+              GtkPrintContext   *ctx,
+              gint               page_nr,
+              char              *filename)
+{
+    cairo_t *cr = NULL;
+    int width = 0, height = 0;
+    unsigned char *data = NULL;
+    int stride = 0;
+
+    width = pdf_get_actions_page (filename, PDF_WIDTH, page_nr);
+    height = pdf_get_actions_page (filename, PDF_HEIGHT, page_nr);
+    cr = gtk_print_context_get_cairo_context (ctx);
+    stride = cairo_format_stride_for_width (CAIRO_FORMAT_ARGB32, width);
+    data = pdf_get_data_page (filename, width, height, page_nr);
+    if (data)
+    {
+      g_autoptr (GdkPixbuf) pix = gdk_pixbuf_new_from_data (data, GDK_COLORSPACE_RGB, TRUE, 8, width, height, stride, NULL, g_free);
+      cairo_scale (cr, (PTS / DPI), (PTS / DPI));
+      gdk_cairo_set_source_pixbuf (cr, pix, 0, 0);
+      cairo_paint (cr);
+      cairo_fill (cr);
+    }
+}
+
+static void
+pdf_end_print (GtkPrintOperation *operation,
+               GtkPrintContext   *context,
+               char              *filename)
+{
+  if (filename)
+  {
+    unlink (filename);
+    g_free (filename);
+  }
+}
+
+static gboolean
+print_pdf(int                    fd,
+          XdpImplPrint         *object,
+          GDBusMethodInvocation *invocation)
+{
+  char *filename = NULL;
+  g_autoptr (GInputStream) istream = NULL;
+  g_autoptr (GOutputStream) ostream = NULL;
+  g_autoptr (GtkPrintSettings) settings = NULL;
+  g_autoptr (GtkPrintOperation) print = NULL;
+  GError *err = NULL;
+  GVariantBuilder opt_builder;
+  int fd2;
+
+  istream = g_unix_input_stream_new (fd, FALSE);
+
+  if ((fd2 = g_file_open_tmp (PACKAGE_NAME "XXXXXX", &filename, &err)) == -1)
+    return FALSE;
+
+  ostream = g_unix_output_stream_new (fd2, TRUE);
+
+  if (g_output_stream_splice (ostream,
+                              istream,
+                              G_OUTPUT_STREAM_SPLICE_NONE,
+                              NULL,
+                              &err) == -1)
+       return FALSE;
+  if (pdf_get_actions_page (filename, PDF_TEST, -1) == 0)
+  {
+    g_free (filename);
+    unlink (filename);
```

Comment:
You can't call `unlink` after `free`, reverse these.

Reply:
Fixed